### DISCUSSION TITLE
Feedback for @cypress/puppeteer Beta

### DISCUSSION THREAD

**Discussion by jennifer-shehane** [cypress-io/cypress#28410] [Questions and Help]
We released a [@cypress/puppeteer](https://github.com/cypress-io/cypress/tree/develop/npm/puppeteer) plugin and we would like to gather feedback on how your experience is using the plugin. We intend to use this feedback to inform what improvements we could make, if any, to the plugin. 

The main intention of the plugin is to allow a way for users to test some use cases of multiple tabs or windows, so that is our main focus when looking at feedback, but we'd be very interested to see other ways that you found it to be useful.

Please include any/all of the following:

- Successes!
- Anything that makes this plugin unusable for your use case.
- Frustrations
- Features that would make this a plugin that you would adopt
- Features that would be nice to have
- Anything else...

**Comment by Benoit-Vasseur** [comment]
Hello,

I cloned the puppeteer folder (https://github.com/cypress-io/cypress/tree/develop/npm/puppeteer) on my local machine (Linux Ubuntu) and the examples do not work :/

I even tried to do a minimal test (just a console.log) and it do not work either :/

```ts
//cypress/e2e/hello.cy.ts
describe('Hello', () => {
    it('hello', () => {
        cy.puppeteer('logHello')
    })
})

// in cypress.config
{
   onMessage: {
       async logHello(browser) {
           return retry(() => {
              console.log('HELLO')
              return 'hello'
            } ) 
     }
}
```

I join my log file 
[cypress_logs.txt](https://github.com/cypress-io/cypress/files/13601184/cypress_logs.txt)

![image](https://github.com/cypress-io/cypress/assets/1382663/ff392277-e994-48f7-bc0e-cf2cf54a9660)

**Reply by chrisbreiding** [reply]
Do you still encounter this issue if you set up a separate project and `npm install --save-dev @cypress/puppeteer` instead of cloning the source?

**Comment by adnanerlansyah403** [comment]
> Hello,
> 
> I cloned the puppeteer folder (https://github.com/cypress-io/cypress/tree/develop/npm/puppeteer) on my local machine (Linux Ubuntu) and the examples do not work :/
> 
> I even tried to do a minimal test (just a console.log) and it do not work either :/
> 
> ```ts
> //cypress/e2e/hello.cy.ts
> describe('Hello', () => {
>     it('hello', () => {
>         cy.puppeteer('logHello')
>     })
> })
> 
> // in cypress.config
> {
>    onMessage: {
>        async logHello(browser) {
>            return retry(() => {
>               console.log('HELLO')
>               return 'hello'
>             } ) 
>      }
> }
> ```
> 
> I join my log file [cypress_logs.txt](https://github.com/cypress-io/cypress/files/13601184/cypress_logs.txt)
> 
> ![image](https://private-user-images.githubusercontent.com/1382663/288786878-ff392277-e994-48f7-bc0e-cf2cf54a9660.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIxMTM0NjIsIm5iZiI6MTcwMjExMzE2MiwicGF0aCI6Ii8xMzgyNjYzLzI4ODc4Njg3OC1mZjM5MjI3Ny1lOTk0LTQ4ZjctYmMwZS1jZjJjZjU0YTk2NjAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMDlUMDkxMjQyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzdmMGIwMmQ3M2YwZDc3YzAwNjExNDA1ZmY3Y2U2M2RiODFlMzk0ZGZhMmNiY2VlNDcxNjQ1NTFiMGVmODIzNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.IHSXGmMTIBdneOq2xJZ4hf67y1CuTunIrGPE3q3pVik)

I got the same issue as like you

**Comment by filiphric** [comment]
I thought a more "Cypress-native" implementation would be much better. Not sure how techincally possible that is but I think building something like `cy.origin()` but for tabs would make much more sense. I’m imagining something like:

```js
cy.visit('/client1') // open first tab

cy.get('input').type('hello world{enter}')

// go to second tab
cy.tab('my second tab', () => {

  cy.visit('/client2')
  cy.get('[data-cy=incoming-message]').should('have.text', 'hello world')

})
```

I think Cypress DX is what people ultimately want. Lack of multi-tab support has a lot of workarounds, but the reason why people stay away from these workarounds is because they don’t want to leave Cypress DX. I haven’t dived too deeply into this, but at first glance it seems like this is not much different from using `cy.task()`. I’m in no way trying to diminish the amount of work that went into this, just trying to look at this from a product/DX perspective.

**Reply by adnanerlansyah403** [reply]
I want this feature to be implemented on Cypress 😭😭😭😭😭😭😭😭😭😭😭😭. Please developer team cypress @jennifer-shehane , make this feature be real 🙏

**Reply by jennifer-shehane** [reply]
@filiphric Thanks for the feedback filip! We did have some initial concepts wrapping a more familiar API into Cypress, with puppeteer doing the background work. For Beta, we took a more incremental approach to implementation - building out the bare bones with a low maintenance cost to us and wait to hear feedback on whether a nicer 'wrapper' would be helpful to users. This is helpful with deciding post-beta actions.

**Comment by wlsf82** [comment]
I’ve tested it in a very [silly project](https://github.com/wlsf82/cypress-puppeteer-plugin-demo), and it works great.

Next, I’ll try a real login scenario that opens a new tab or window.

Thanks, Cypress team!

**Comment by wlsf82** [comment]
@jennifer-shehane, although the [documentation](https://github.com/cypress-io/cypress/tree/develop/npm/puppeteer#compatibility) mentions the compatibility with Electron, [in a project I tried](https://github.com/wlsf82/cypress-puppeteer-plugin-demo/tree/electron-bug), it only works with Chromium web-based browsers (e.g., Chrome and Edge), [not on Electron](https://github.com/wlsf82/cypress-puppeteer-plugin-demo/commit/5872cff31e1b708a56b1a3b555ba04467c5861b1).

See [this CI failure](https://github.com/wlsf82/cypress-puppeteer-plugin-demo/actions/runs/7214984216/job/19658283733) as a reference.

And [here's](https://cloud.cypress.io/projects/u8w8pa/runs/12/overview?roarHideRunsWithDiffGroupsAndTags=1) the same run on the Cypress Cloud.

To prevent that, [I'm filtering for only chromium web-based browsers](https://github.com/wlsf82/cypress-puppeteer-plugin-demo/blob/main/cypress.config.js#L12).

**Reply by jennifer-shehane** [reply]
@wlsf82 Thanks for reporting! We'll take a look at this. May be a bit slower with the end of the year coming up.

**Comment by fabiojfadsg** [comment]
@jennifer-shehane , I created a test scenario where I believe this plugin would be essential.
Scenario: Simulation of a chat conversation with 2 users with different profiles. 
Analyzing the plugin I saw that the second tab opens as expected but we have some problems when we add tests to the input fields:
1- Texts are being typed slowly, character by character with an average time of one second for each character.
2- When we click to go to the next step, it appears that the second tab displays the text entered, but shows the message that it is necessary to enter a name, appearing as if there was no typing.
 
I left a folder in the project called “evidence”, containing a video and images of the problem. Link para projeto 
[Github](https://github.com/fabiojfadsg/cypress-puppeteer)

**Reply by jennifer-shehane** [reply]
@fabiojfadsg For the delay in between typing, you have passed a `delay` option to the Puppeteer type method, this is the "delay which is the Time to wait between key presses in milliseconds." So you have added an artificual 1 sec delay in between each character press, you can edit that in your tests. https://pptr.dev/api/puppeteer.page.type

For the second problem, it looks like it did submit the username and password, but it displaying an error about an invalid username. And it clears the username and password field when hitting the submit button, as shown in the gif below I recreated outside of Cypress.

![](http://g.recordit.co/a7tnOXMCz8.gif)

**Comment by mforzano** [comment]
I'm encountering a weird issue with trying to use the latest version (0.1.4). I'm essentially following the "Switching to a new tab" example in the readme, but without actually switching tabs and simply logging the page title, URL and content of the current page to the console. But, instead of the page under test being loaded in Puppeteer, I'm getting the Cypress test runner page.

So for example:
1. I load a URL, i.e. https://www.amazon.com using cy.visit
2. I invoke my Cypress task.
3. The following is logged:

```
Page title is: MyCypressSample, URL: https://www.amazon.com/__/#/specs/runner?file=cypress/e2e/sample-script-open-homepage.cy.ts
```

Where 'MyCypressSample' is the name of my test pacakge.

Also, if I I log the length of the pages array I get 1.

I would expect that Puppeteer has access to the page under test as opened by cy.visit.

Any pointers would be much appreciated.

**Comment by Alfy102** [comment]
Hi, I am trying to use this plugin to enable tracing. I am getting `Cannot read properties of undefined (reading 'tracing')`. 
I have `cypress-esbuild-preprocessor` and `cypress-cucumber-preprocessor`.

```
beforeEach(() => {
  cy.puppeteer('traceStarts');
});


afterEach(() => {
  cy.puppeteer('traceEnds');
});
```

```
      setup({
        on,
        onMessage: {
          async traceStarts(browser: PuppeteerBrowser, page: Page) {
            await page.tracing.start({ path: 'trace.json' });
            return null;
          },

          async traceEnds(browser: PuppeteerBrowser, page: Page) {
            await page.tracing.stop();
            return null;
          }
        },
      });
```
![image](https://github.com/cypress-io/cypress/assets/8023501/2669a8fa-c8b8-4001-98c6-43f966c281c6)

**Comment by annamazurek-qa** [comment]
Hello!

I have problems with types of the Browser. Browser property is passed by cy.puppeteer(); and the type of the browser they pass doesn't match with the official Browser type. The reason is:
"Type 'Browser' is missing the following properties from type 'Browser': createBrowserContext, #privatets(2352)"

Could you fix it somehow?

Thanks!

**Comment by askhanna-mdsol** [comment]
CypressError: {cy.task('__cypressPuppeteer__')} timed out after waiting {60000ms} I am now observing timeout failures in GIThub actions, even if it functions properly on my local system.
Has anyone else have similar experiences?

Debug logs
```
CypressError: `cy.task('__cypressPuppeteer__')` timed out after waiting `60000ms`.\n" + '\n' + '[https://on.cypress.io/api/task\n](https://on.cypress.io/api/task/n)' + ' at <unknown> ([**************************/__cypress/runner/cypress_runner.js:134985:70](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:134985:70))\n' + ' at tryCatcher ([**************************/__cypress/runner/cypress_runner.js:1807:23](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:1807:23))\n' + ' at <unknown> ([**************************/__cypress/runner/cypress_runner.js:4186:41](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:4186:41))\n' + ' at tryCatcher ([**************************/__cypress/runner/cypress_runner.js:1807:23](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:1807:23))\n' + ' at Promise._settlePromiseFromHandler ([**************************/__cypress/runner/cypress_runner.js:1519:31](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:1519:31))\n' + ' at Promise._settlePromise ([**************************/__cypress/runner/cypress_runner.js:1576:18](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:1576:18))\n' + ' at Promise._settlePromise0 ([**************************/__cypress/runner/cypress_runner.js:1621:10](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:1621:10))\n' + ' at Promise._settlePromises ([**************************/__cypress/runner/cypress_runner.js:1697:18](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:1697:18))\n' + ' at _drainQueueStep ([**************************/__cypress/runner/cypress_runner.js:2407:12](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:2407:12))\n' + ' at _drainQueue ([**************************/__cypress/runner/cypress_runner.js:2400:9](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:2400:9))\n' + ' at Async._drainQueues ([**************************/__cypress/runner/cypress_runner.js:2416:5](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:2416:5))\n' + ' at Async.drainQueues ([**************************/__cypress/runner/cypress_runner.js:2286:14](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:2286:14))\n' + 'From Your Spec Code:\n' + ' at Context.eval ([**************************/__cypress/tests?p=cypress/e2e/ui/SiteMonitoring/Features/M3_ConductVisit_SE.feature:32035:12](https://login-sandbox.imedidata.net/__cypress/tests?p=cypress/e2e/ui/SiteMonitoring/Features/M3_ConductVisit_SE.feature:32035:12)
```

**Reply by askhanna-mdsol** [reply]
> CypressError: {cy.task('**cypressPuppeteer**')} timed out after waiting {60000ms} I am now observing timeout failures in GIThub actions, even if it functions properly on my local system. Has anyone else have similar experiences?
> 
> Debug logs CypressError: `cy.task('__cypressPuppeteer__')` timed out after waiting `60000ms`.\n" + '\n' + '[https://on.cypress.io/api/task\n](https://on.cypress.io/api/task/n)' + ' at ([**************************/__cypress/runner/cypress_runner.js:134985:70](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:134985:70))\n' + ' at tryCatcher ([**************************/__cypress/runner/cypress_runner.js:1807:23](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:1807:23))\n' + ' at ([**************************/__cypress/runner/cypress_runner.js:4186:41](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:4186:41))\n' + ' at tryCatcher ([**************************/__cypress/runner/cypress_runner.js:1807:23](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:1807:23))\n' + ' at Promise._settlePromiseFromHandler ([**************************/__cypress/runner/cypress_runner.js:1519:31](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:1519:31))\n' + ' at Promise._settlePromise ([**************************/__cypress/runner/cypress_runner.js:1576:18](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:1576:18))\n' + ' at Promise._settlePromise0 ([**************************/__cypress/runner/cypress_runner.js:1621:10](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:1621:10))\n' + ' at Promise._settlePromises ([**************************/__cypress/runner/cypress_runner.js:1697:18](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:1697:18))\n' + ' at _drainQueueStep ([**************************/__cypress/runner/cypress_runner.js:2407:12](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:2407:12))\n' + ' at _drainQueue ([**************************/__cypress/runner/cypress_runner.js:2400:9](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:2400:9))\n' + ' at Async._drainQueues ([**************************/__cypress/runner/cypress_runner.js:2416:5](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:2416:5))\n' + ' at Async.drainQueues ([**************************/__cypress/runner/cypress_runner.js:2286:14](https://login-sandbox.imedidata.net/__cypress/runner/cypress_runner.js:2286:14))\n' + 'From Your Spec Code:\n' + ' at Context.eval ([**************************/__cypress/tests?p=cypress/e2e/ui/SiteMonitoring/Features/M3_ConductVisit_SE.feature:32035:12](https://login-sandbox.imedidata.net/__cypress/tests?p=cypress/e2e/ui/SiteMonitoring/Features/M3_ConductVisit_SE.feature:32035:12)

Updating the base Cypress image fixed this, yet occasionally I encounter the error below.

cy.puppeteer() failed with the following error:
) Cannot communicate with the Cypress Chrome extension. Ensure the extension is enabled when using the Puppeteer plugin.

**Comment by ssomanathan-mdsol** [comment]
@jennifer-shehane Iam trying to use cypress-puppeteer to take screenshots as the Cypress screenshots have an issue in taking fullPage of our app, with part of the page showing up as duplicated in the screenshot taken. I could see multiple issues have been raised on this already.

With cypress-puppeteer, I could take the screenshots, but is there a way i could pass from Cypress which part of the browser that the screenshot needs to be taken ? Currently puppeteer takes a screenshot of the whole of the browser which includes the cypress run log on the left . 
The current workaround i have done is to remove the runlog from the DOM, take the screenshot and re-attach the runlog. 
But there would be a better way to do this . If there is , please could you let me know .. 
Many Thanks..

**Reply by aleksandrlat** [reply]
Maybe try `-no-runner-ui` option https://docs.cypress.io/app/references/command-line#cypress-run-no-runner-ui

Also puppeteer allows to make screenshot of element
https://pptr.dev/api/puppeteer.elementhandle.screenshot
Test runs in iframe with class `aut-iframe`. Maybe something like this
```
await page.locator('iframe.aut-iframe').screenshot()
```

**Comment by aleksandrlat** [comment]
Why puppeteer and not playwright? I see playwright is already used for webkit automation
And CDP is used directly too

**Comment by alperkuertuel** [comment]
@jennifer-shehane:

In my current testing setup, I am using Cypress 14, puppeteer-core 24.1.0, and @cypress/puppeteer v0.1.5. I encountered an issue when trying to use the Firefox browser with Cypress and the Puppeteer plugin. The error message states:

```
cy.puppeteer() failed with the following error: > Only browsers in the "Chromium" family are supported. You are currently running a browser with the family: firefox
```

I am aware that puppeteer-core has supported Firefox since version 23. Given this, I was expecting that integration with Firefox would also be possible using @cypress/puppeteer.

Could you provide some guidance on whether there is a known limitation or plan to support Firefox in @cypress/puppeteer? Should I open a new issue with a detailed description for this beta plugin, or is there an alternative workaround to enable Firefox support with the current versions?

Thank you for your assistance!

**Comment by Jasperhino** [comment]
Hi, in our big enterprise codebase we have mocked the whole backend requests using cypress interceptors. Now I want to create a test that test the syncing of operations between multiple tabs in the browser. I thought this is the perfect usecase to try the pupeteer plugin for cypress. I got started, setup the first test the application opens a new puppeteer tab like expected and fills the login information. But now the request interceptors we coded in cypress don't apply to the pupeteer tab. Is there a solution to reuse these in the puppeteer tab? Or something else I have not thought about yet?

Would be a shame if we had to rewrite our whole mocking just for this one test case :/

Best, Jasper

**Comment by spdaley** [comment]
We tried this out to see if it solved some of our multi window testing issues and it seems to work well. My one issue is that it seems like it has not been updated in awhile so some of its third-party dependencies fail some security audits. In particular we get these now when trying to use this:

```
GHSA-vj76-c3g6-qr5v tar-fs (3.0.4)
GHSA-8cj5-5rvv-wf4v tar-fs (3.0.4)
GHSA-pq67-2wwv-3xjx tar-fs (3.0.4)
GHSA-3h5v-q93c-6h6q ws (8.16.0)
GHSA-58qx-3vcg-4xpx ws (8.16.0)
GHSA-96hv-2xvq-fx4p ws (8.16.0)
```

I believe these all come from the inclusion of puppeteer-core which is at v21 and a few years old. Their latest is v25. Is there any intent to update this plugin at some point?
